### PR TITLE
Webhook retry enhancement docs

### PR DESCRIPTION
**Pull request by shikhartanwar**
## Description
This PR makes the following changes -
 - Adding docs incorporating webhook retry improvements
    - Change in retry policy (retried upto 24h now)
    - Introduction of unstable status
    - Captured corner-case/outlier-scenario when an event is dropped
  - Re-emphasized importance of signature verification for webhook deliveries

## Note
PR raised against which branch -
 - [x] master (Production)
 - [ ] stage (Stage)

### CODE REVIEW

**Review by rperelma — APPROVED** [review]
LGTM

File: intro/webhooks_intro.md [hunk 1]
```diff
@@ -193,27 +193,30 @@ The *Status* of the registration should show as **Active**. If the registration
...
-Adobe marks the webhook as invalid and stops sending requests. 
+If `Adobe I/O Events` fails to receive a successful response code from your webhook within 10 seconds, it retries the request, including a special header `x-adobe-retry-count` (this header indicates how many times the delivery of an event or a batch of events has been attempted).
 
-To restart the flow of requests, once you have fixed the problem preventing your webhook from responding, 
-you must log into the `Adobe Developer Console`, edit your events registration, 
-it will re-trigger a webhook challenge request, and eventually a webhook re-activation.
+`Adobe I/O Events` will keep on retrying delivery to your webhook for **24 hours** using exponential and fixed backoff strategies. The first retry is attempted after 1 minute and the period between retries doubles after each attempt (second retry is after 2m, etc.), but is at most 15 minutes.
+
+If an event isn't delivered after 2 hours of retries, `Adobe I/O Events` marks the webhook as **Unstable**, but still keeps on attempting delivery. This gives sufficient time to fix whatever problems your are facing and avoid your webhook from getting marked as Disabled.
```
Comment (sgera) [inline intro/webhooks_intro.md:208]: nit: your are facing -> you are facing

Or rephrase it to "This gives you sufficient time to restore your webhook"?

File: intro/webhooks_intro.md [hunk 2]
```diff
@@ -193,27 +193,30 @@ The *Status* of the registration should show as **Active**. If the registration
...
+
+If all retry attempts get exhausted and the event still isn't delivered (webhook not responding or responding with a non `2XX` response), `Adobe I/O Events` drops the events, marks the webhook as **Disabled**, and stops sending any further events.
```
Comment (sgera) [inline intro/webhooks_intro.md:210]: "marks the webhook as **Disabled**" OR "marks the event-registration as **Disabled**"? I notice that we are using 'webhook' instead of 'event-registration'. Is it to remain consistent with the older version of this doc?

Comment (sgera) [reply intro/webhooks_intro.md:210]: Here we are using "event-registration" - https://github.com/AdobeDocs/adobeio-events/pull/87/commits/a0d1856e68b8ddba5f3f4300811836280683c8db#diff-7ba92183051d83f15ba331c96cdaa686e48ec16f9e97c46a30f3789027c3dfdbR96

File: support/faq.md [hunk 3]
```diff
@@ -93,19 +93,27 @@ Yes:
...
 
-Note you can then use the Journaling API to **retrieve** events that were published while your webhook was down and once the webhook is back up, you can re-enable it, see the question below.
+`Adobe I/O Events` will keep on retrying delivery to your webhook for **24 hours** using exponential and fixed backoff strategies. The first retry is attempted after 1 minute and the period between retries doubles after each attempt, but is at most 15 minutes (see below table outlining the retry pattern).
+
+| Retry Attempt        | 1  | 2  | 3  | 4  | 5   | 6   | 7   | ... |
+|----------------------|----|----|----|----|-----|-----|-----|-----|
+| Retry After Interval | 1m | 2m | 4m | 8m | 15m | 15m | 15m | ... |
+
+If an event isn't delivered after 2 hours of retries, `Adobe I/O Events` marks the webhook as **Unstable**, but still keeps on attempting delivery. This gives sufficient time to fix whatever problems your are facing and avoid your webhook from getting marked as Disabled.
```
Comment (sgera) [inline support/faq.md:106]: https://github.com/AdobeDocs/adobeio-events/pull/87/commits/a0d1856e68b8ddba5f3f4300811836280683c8db#r748016091

**Review by sgera — COMMENTED** [review]
Added a few minor comments.

File: intro/webhooks_intro.md (hunk 2 above)
Comment (shikhartanwar) [reply intro/webhooks_intro.md:210]: Hmm good point. Shall we change this throughout with `webhook-registration`?

File: intro/webhooks_intro.md [hunk 4]
```diff
@@ -193,27 +195,30 @@ The *Status* of the registration should show as **Active**. If the registration
 
 ![Event Registration Details tab in Adobe Developer Console](../img/events-registration-details.png)
 
-### Troubleshooting a Disabled Registration Status
+### Troubleshooting Unstable/Disabled Registration Status
```
Comment (sgera) [inline intro/webhooks_intro.md:198]: Should we provide details for this state transition - When does an event-registration go from "Unstable or Disabled" to "Active/Stable/Healthy"?

File: intro/webhooks_intro.md (hunk 1 above)
Comment (shikhartanwar) [reply intro/webhooks_intro.md:208]: Thank you for catching this. Rephrased now.

File: support/faq.md (hunk 3 above)
Comment (shikhartanwar) [reply support/faq.md:106]: Done

File: intro/webhooks_intro.md (hunk 4 above)
Comment (shikhartanwar) [reply intro/webhooks_intro.md:198]: Do you think it will add value? If I think from a developer perspective, I would be concerned as to why my webhook registration is marked as `Unstable`/`Disabled` (which is covered in the faq doc). If it shows up as `Active` then I may not worry what the previous state was.

Also, transition from a `Disabled to Active` state requires explicit action from the developer (which is covered in the faq doc). However, transition from `Unstable` to `Active` state is implicit as long as the webhook becomes reachable again.

I would like to keep things simple but if you feel otherwise, let me know.

File: intro/webhooks_intro.md (hunk 2 above)
Comment (sgera) [reply intro/webhooks_intro.md:210]: On the Developer Console, we use `event-registration`. I think we should remain consistent and use that only.

File: intro/webhooks_intro.md [hunk 5]
```diff
@@ -193,27 +195,30 @@ The *Status* of the registration should show as **Active**. If the registration
...
-Adobe marks the webhook as invalid and stops sending requests. 
+If `Adobe I/O Events` fails to receive a successful response code from your webhook within 10 seconds, it retries the request, including a special header `x-adobe-retry-count` (this header indicates how many times the delivery of an event or a batch of events has been attempted).
 
-To restart the flow of requests, once you have fixed the problem preventing your webhook from responding, 
-you must log into the `Adobe Developer Console`, edit your events registration, 
-it will re-trigger a webhook challenge request, and eventually a webhook re-activation.
+`Adobe I/O Events` will keep on retrying delivery to your webhook for **24 hours** using exponential and fixed backoff strategies. The first retry is attempted after 1 minute and the period between retries doubles after each attempt (second retry is after 2m, etc.), but is at most 15 minutes.
+
+If an event isn't delivered after 2 hours of retries, `Adobe I/O Events` marks the webhook as **Unstable**, but still keeps on attempting delivery. This gives you sufficient time to restore your webhook, and avoid it from getting marked as Disabled.
```
Comment (sgera) [inline intro/webhooks_intro.md:208]: To cover for `Unstable` to `Active` transition, we can append the following line in this paragraph -
```
Once the webhook is restored, `Adobe I/O Events` marks the webhook-registration to **Active** state on the next successful event delivery.
```

File: intro/webhooks_intro.md (hunk 4 above)
Comment (sgera) [reply intro/webhooks_intro.md:198]: The only transition left unanswered is from `Unstable` to `Active`. See if this looks good to you - https://github.com/AdobeDocs/adobeio-events/pull/87/files#r748027235. If not, then we can leave it as is.

File: intro/webhooks_intro.md (hunk 5 above)
Comment (shikhartanwar) [reply intro/webhooks_intro.md:208]: Sure, I can add that.

File: intro/webhooks_intro.md (hunk 2 above)
Comment (shikhartanwar) [reply intro/webhooks_intro.md:210]: My only concern is that throughout this doc, we have been using `webhook` heavily. Changing it to `webhook-registration` would be a less transparent change and still convey what we are trying to say.

Comment (sgera) [reply intro/webhooks_intro.md:210]: Introducing a new term `webhook-registration` is not preferable. Either we continue to use `webhook` (and leave this inconsistency to be handled in the future), or we become consistent by using `event registration` everywhere.

File: intro/webhooks_intro.md (hunk 5 above)
Comment (sgera) [reply intro/webhooks_intro.md:208]: `webhook-registration` -> `event registration`

Comment (shikhartanwar) [reply intro/webhooks_intro.md]: Done

File: intro/webhooks_intro.md (hunk 2 above)
Comment (shikhartanwar) [reply intro/webhooks_intro.md:210]: After discussion with @sgera, `webhook-registration` will be yet another new term, so it would be good if we use `event-registration` throughout.

I'll make those changes.

Comment (shikhartanwar) [reply intro/webhooks_intro.md]: Changed to `event registration` throughout based on the below console screenshot:

event-registration

File: intro/webhooks_intro.md (hunk 4 above)
Comment (shikhartanwar) [reply intro/webhooks_intro.md:198]: I added that. See https://github.com/AdobeDocs/adobeio-events/pull/87/commits/060427dec8a16358a026713555f8e15a9a386e82.

File: intro/webhooks_intro.md [hunk 6]
```diff
@@ -8,7 +8,7 @@ With the right webhook in place, your application is instantly notified that thi
 
 Please refer to the `Adobe Developer Console` documentation on how to [Add Events to a project](https://www.adobe.io/apis/experienceplatform/console/docs.html#!AdobeDocs/adobeio-console/master/services-add-event.md)
 
-To start receiving events, you register a webhook, specifying a webhook URL and the types of events you want to receive. Each event will result in a HTTP request to the given URL, notifying your application. This guide provides an introduction to webhooks, including:
+To start receiving events, you register an event registration, specifying a webhook URL and the types of events you want to receive. Each event will result in a HTTP request to the given URL, notifying your application. This guide provides an introduction to webhooks, including:
```
Comment (sgera) [inline intro/webhooks_intro.md:11]: `To start receiving events, you register an event registration` -> `To start receiving events, you create an event registration`?

File: intro/webhooks_intro.md (hunk 2 above)
Comment (sgera) [reply intro/webhooks_intro.md]: Looks really good, thank you!

File: intro/webhooks_intro.md (hunk 6 above)
Comment (sangeetha5491) [reply intro/webhooks_intro.md:11]: specifying a webhook URL -> **specify** a webhook URL

Comment (shikhartanwar) [reply intro/webhooks_intro.md]: Yes, better. Done.

File: intro/webhooks_intro.md (hunk 2 above)
Comment (shikhartanwar) [reply intro/webhooks_intro.md]: The [SDK webhook page](https://www.adobe.io/apis/experienceplatform/events/docs.html#!adobedocs/adobeio-events/master/sdk/sdk_webhooks.md) is full of `webhook registration` terms, so we can't get rid of it. It's there in our SDK methods as well. However, I have updated the intro section to align with the terminology. See https://github.com/AdobeDocs/adobeio-events/pull/87/commits/8f1c0147ef7215e91d0a2bc6694790589312ab48.

**Review by sgera — APPROVED** [review]

File: intro/webhooks_intro.md [hunk 7]
```diff
@@ -8,28 +8,30 @@ With the right webhook in place, your application is instantly notified that thi
 
 Please refer to the `Adobe Developer Console` documentation on how to [Add Events to a project](https://www.adobe.io/apis/experienceplatform/console/docs.html#!AdobeDocs/adobeio-console/master/services-add-event.md)
 
-To start receiving events, you register a webhook, specifying a webhook URL and the types of events you want to receive. Each event will result in a HTTP request to the given URL, notifying your application. This guide provides an introduction to webhooks, including:
+To start receiving events, you create an event registration, specifying a webhook URL and the types of events you want to receive. Each event will result in a HTTP request to the given URL, notifying your application. This guide provides an introduction to webhooks, including:
```
Comment (sangeetha5491) [inline intro/webhooks_intro.md:11]: specifying a webhook URL -> **specify** a webhook URL

File: support/faq.md [hunk 8]
```diff
@@ -97,57 +94,50 @@ Yes:
...
-their IPs change over time.
-
-Reminder: each Adobe I/O Events HTTP request
-holds an signature header (see the previous question), however if this is not enough 
-and if the above is a non-negotiable requirement, 
-you may choose to use the pull model instead, 
-and leverage our [Journaling API](../intro/journaling_intro.md).          
+The answer is **no**. Adobe I/O Events notifications services are hosted on AWS and Azure, their IPs change over time.
 
+*Reminder*: Each Adobe I/O Events HTTP request holds an signature header (see the previous question), however if this is not enough and if the above is a non-negotiable requirement, you may choose to use the pull model instead, and leverage our [Journaling API](../intro/journaling_intro.md).
```
Comment (sangeetha5491) [inline support/faq.md:134]: an signature header -> "a" signature header

**Review by sangeetha5491 — COMMENTED** [review]
very minor comments

File: intro/webhooks_intro.md (hunk 7 above)
Comment (shikhartanwar) [reply intro/webhooks_intro.md:11]: I think specifying is fine. I'll get rid of the comma before that.

File: support/faq.md (hunk 8 above)
Comment (shikhartanwar) [reply support/faq.md]: Done

File: intro/webhooks_intro.md (hunk 7 above)
Comment (shikhartanwar) [reply intro/webhooks_intro.md]: Changed to `To start receiving events, you create an event registration specifying a webhook URL and the types of events you want to receive.`
